### ISSUE TITLE
another unexpected error with .bank directive

### ISSUE THREAD

**Issue by mrdudz** [cc65/cc65#1538]
minimal example:
```
.define BASEADDRESS         $123456
.define BANKNR              $12

; this works always
;offset = (BASEADDRESS + (BANKNR * $1000))

; none of these work with the second LDA
offset = (BASEADDRESS + ((<.bank(banksym)) * $1000))
;offset = (BASEADDRESS + (<(.bank(banksym) * $1000)))
;offset = (BASEADDRESS + ((.bank(banksym) * $1000)&$ffff))

banksym: 

        lda #<(offset >> 0) & $ff       ; working

        lda #(offset >> 0) & $ff        ; range error ?
```
in the real code, offset will be a 25 bit address. there are 4 LDA like above shifting by 0/8/16/24 and then masking with $ff. this works fine with the first expression that does not use .bank - as soon as i use .bank in it, i need to prefix the shift with a < operator as in the example, else i get a range error. i dont quite understand why, as the & $ff should reduce the expression to 8 bit either way. i'd expect both to work, regardless of the .bank directive being used.

**Comment by mrdudz** [comment]
excuse me, how is this not a bug?

**Comment by greg-king5** [comment]
In some ways, ca65 is a strange animal.  It's more like a compiler than other assemblers are.  Expressions carry a data width.

ca65 cannot know what value `.bank()` will have -- it must pass the entire expression to ld65.  It doesn't know what the final width will be, but it does make some guesses.  It sees that `BASEADDRESS` is too wide.  It sees that `$1000` is too wide.
It can't know what `"& $ff"` will do.  It does know what `<` will do.

Therefore, ca65 guesses that the first `lda` won't be too wide.  It guesses that the second one will be too wide.

**Comment by mrdudz** [comment]
> It can't know what "& $ff" will do. It does know what < will do.

what? it cant know that whatever ANDed with $ff is 8bits? you must be kidding. And it very well does - because with
```
offset = (BASEADDRESS + (BANKNR * $1000))
lda #(offset >> 0)
```
you do also get the out of range error. thats why the & $ff was there in the first place.

yes, i realize that the difference is that one is a constant expression and the other is not - however, there is no reason it shouldnt be able to determine that the result is 8bit in either case.

definetly a bug to me.

**Comment by spiro-trikaliotis** [comment]
This is general behaviour of ca65 for some years now. It was this way since a long time. There was even a discussion with Uz on the old mailing list about this behaviour. See, for example, his posting from Wed, 16 Apr 2014 16:15:50 +0200 (Message-ID: <20140416141550.GA14810@trixie.musoftware.de>). Or the statements that "bitwise or is broken" (Simon Scott, Sat, 24 Dec 2011 19:51:21 +0800) because, for example, AND #~XYZ with XYZ being an 8 bit value does not work.

Uz also gave a reasoning of this behaviour. It was deliberately introduced not long before that.

Note, also, that you yourself answered in this thread back then. You said that it might be a reason for you to switch the toolchain (Fri, 13 Jan 2012 17:18:55 +0100).

One can argue if this is a bug or not. Anyway, it is clearly a problem: The old behaviour let you write wrong code (for example, lda #-300). The fix for that problem introduced a new problem.

We could try to find out if we find a better behaviour here, though.

**Comment by mrdudz** [comment]
this is NOT about the < being required. i can accept this. it is purely about .bank not working *eventhough* the & $ff is there. those are two different issues. the later can (and should, imho) be fixed, it just needs some more/better expression evaluating/checking in the linker instead of rejecting things early in the assembler.

(mentioning almost 10 year old mails without actually linking to them is pretty useless to me, i am not going to the attic and check my old pc =P)

**Comment by mrdudz** [comment]
> it just needs some more/better expression evaluating/checking in the linker instead of rejecting things early in the assembler.

actually this particular case should be fixable the way i mentioned, right in the assembler. even more of a reason for why it should be :)

**Comment by oliverschmidt** [comment]
> mentioning almost 10 year old mails without actually linking to them is pretty useless to me, i am not going to the attic and check my old pc =P

The content in question is "at your fingertips": https://cc65.github.io/mailing-lists.html#musoftware

**Comment by mrdudz** [comment]
> > mentioning almost 10 year old mails without actually linking to them is pretty useless to me, i am not going to the attic and check my old pc =P
> 
> The content in question is "at your fingertips": https://cc65.github.io/mailing-lists.html#musoftware

I'd rather invest the time in looking at expr.c than digging up 10 year old mails :)

**Comment by mrdudz** [comment]
Ok i skimmed a bit over expr.c and i think it can actually be implemented cleanly (although needs a bit of work). Before i actually start doing it, let me reiterate:

The issue is _not_ that certain operations generally evaluate to a result of (i think) 24 bit, ie the example given by spiro 
```
lda #~$12
```
- This will not work and due to how ca65 works must be written as either
```
lda #<(~$12)
```
or
```
lda #(~$12) & $ff
```
Again, this is _not_ the problem i am trying to address - this is fine, this is how ca65 works.

The Problem i am trying to tackle is when a non constant part of an expression marks the entire expression as "unknown" (actually it is max size, 24 bit) size, eventhough even without knowing the non constant part, the expression size _can_ be evaluated. ie
```
lda #<(.bank(foo))
```
and
```
lda #(.bank(foo)) & $ff
```
should both work, because both can be safely assumed 8 bit (there is actually no difference between the two, except they use a different operator)

To do this, we need some extra checking in expr.c for the bitwise and operator. As in: when one side of the bitwise and is max. 8bit, then mark the whole expression as such. When its 16bit max, mark the expression as 16bit, etc. This should be straightforward and have no side effects whatsoever (unless i missed something, please speak up). The various operators that deal with hi/lo bytes/words already implement this sort of thing. 

Once this works for bitwise and we can perhaps do it for other operators too (although i am not sure which that would be, and what the side effect could be - but one thing at a time).

So, before i go on and put significant time into it - do we want this, and if not what speaks against it?

**Comment by mrdudz** [comment]
So, comments? I am very tempted to tag this as bug again at least - because unless someone can tell a case where the above mentioned fix creates an unwanted side effect, this pretty much is a bug.

**Comment by oliverschmidt** [comment]
The thing I know fore sure is that I don't like the only two team members (in contrast to contributors) to discuss if an issue is a question or a bug :-(

> So, before i go on and put significant time into it - do we want this, and if not what speaks against it?

My perspective:

- I believe that in contrast to the C compiler the risk of getting into a big mess of regressions with the assembler is  pretty small. So I'm from the pure organizational perspective pro a change.
- Regarding the actual topic I must admit that I don't understand where @greg-king5's and your perspective differ. And I currently just don't have the time budget to look into deep enough to gain my own perspective.

**Comment by mrdudz** [comment]
> I must admit that I don't understand where @greg-king5's and your perspective differ. 

I think he misunderstood what i was actually trying to address, just like spiro did. As explained above, this is NOT about the case when an expression somewhat unexpected turns into 24bit (like ~$00 turns into $ffffff). The only thing i want to do is make the expression evulation aware of the fact that ANDing two values always results in a value whose number of bits is limited by the operand with the least number of bits - and that even when one of the operands is not constant.

**Comment by oliverschmidt** [comment]
If you think you can add the behavior desired by you without overly complicating the code I'm fine with merging a PR doing so.

**Comment by greg-king5** [comment]
No objection from me.

**Comment by oliverschmidt** [comment]
@greg-king5: Thanks for the feedback :-)

**Comment by mrdudz** [comment]
> If you think you can add the behavior desired by you without overly complicating the code I'm fine with merging a PR doing so.

As said, it should be doable cleanly, all the bits and pieces are already there, just needs some special case handling for bitwise and - i will look at it, will take a while, i guess :)

**Comment by oliverschmidt** [comment]
Great!

**Comment by mrdudz** [comment]
closing, fixed in #1724, testcase added
